Framework/Algorithms/src/ConvertSpectrumAxis.cpp

```diff
@@ -95,15 +96,16 @@ void ConvertSpectrumAxis::exec() {
       emode = 2;
     const double delta = 0.0;
     double efixed;
+    auto spectrumInfo = inputWS->spectrumInfo();
     for (size_t i = 0; i < nHist; i++) {
       std::vector<double> xval{inputWS->x(i).front(), inputWS->x(i).back()};
-      IDetector_const_sptr detector = inputWS->getDetector(i);
       double twoTheta, l1val, l2;
-      if (!detector->isMonitor()) {
-        twoTheta = inputWS->detectorTwoTheta(*detector);
-        l2 = detector->getDistance(*sample);
+      if (!spectrumInfo.isMonitor(i)) {
+        twoTheta = spectrumInfo.twoTheta(i);
+        l2 = spectrumInfo.l2(i);
         l1val = l1;
-        efixed = getEfixed(detector, inputWS, emode); // get efixed
+        efixed =
+            getEfixed(spectrumInfo.detector(i), inputWS, emode); // get efixed
```

Comment:
I have looked several times into having a nicer way to get `EFixed` and `EMode`. The problem is that Mantid has two ways of storing/getting it (instrument parameter or in `Run`, I think). So this is not easy to refactor. Also, there is some weird behavior with the way parameters are obtain that might make trouble. For example, if you store a parameter on your detectors, and then group detectors into a spectrum, the parameters will be ignored when getting parameters for the spectrum.

I agree 100% that this should be done in a better way, but it seems to be a can of worms, so we have to think carefully when and how we can do it.